src/sinks/gcp_stackdriver_logging.rs

```diff
@@ -0,0 +1,320 @@
+    pub auth: GcpAuthConfig,
+
+    #[serde(default, flatten)]
+    pub batch: BatchBytesConfig,
+    #[serde(flatten)]
+    pub request: TowerRequestConfig,
+
+    pub tls: Option<TlsOptions>,
+}
+
+#[derive(Clone, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+pub enum StackdriverLogName {
+    #[serde(rename = "billing_account_id")]
+    BillingAccount(String),
+    #[serde(rename = "folder_id")]
+    Folder(String),
+    #[serde(rename = "organization_id")]
+    Organization(String),
+    #[derivative(Default)]
+    #[serde(rename = "project_id")]
+    Project(String),
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+pub struct StackdriverResource {
+    #[serde(rename = "type")]
+    pub _type: String,
+    #[serde(flatten)]
+    pub labels: HashMap<String, String>,
+}
+
+inventory::submit! {
+    SinkDescription::new::<StackdriverConfig>("gcp_stackdriver_logging")
+}
+
+lazy_static! {
+    static ref REQUEST_DEFAULTS: TowerRequestConfig = TowerRequestConfig {
+        rate_limit_num: Some(1000),
+        rate_limit_duration_secs: Some(1),
+        ..Default::default()
+    };
+    static ref URI: Uri = "https://logging.googleapis.com/v2/entries:write"
+        .parse()
+        .unwrap();
+}
+
+#[typetag::serde(name = "gcp_stackdriver_logging")]
+impl SinkConfig for StackdriverConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        let creds = self.auth.make_credentials(Scope::LoggingWrite)?;
+        let sink = self.service(&cx, &creds)?;
+        let healthcheck = self.healthcheck(&cx, &creds)?;
+
+        Ok((sink, healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "gcp_stackdriver_logging"
+    }
+}
+
+impl StackdriverConfig {
+    fn service(
+        &self,
+        cx: &SinkContext,
+        creds: &Option<GcpCredentials>,
+    ) -> crate::Result<super::RouterSink> {
+        let batch = self.batch.unwrap_or(bytesize::kib(5000u64), 1);
+        let request = self.request.unwrap_with(&REQUEST_DEFAULTS);
+
+        let tls_settings = TlsSettings::from_options(&self.tls)?;
+        let creds = creds.clone();
+
+        // We need to cap the maximum length of the encoded request, so
+        // we encode each event into JSON separately and then splice the
+        // result into a relatively short wrapper.
+        let wrapper = self.write_request();
+        let wrapper_splice = wrapper
+            .find(SPLICE_MAGIC)
+            .expect("Unexpected encoded wrapper format")
+            + SPLICE_OFFSET;
+
+        let http_service = HttpService::builder(cx.resolver())
+            .tls_settings(tls_settings)
+            .build(move |mut logs: Vec<u8>| {
+                logs.pop(); // Strip the trailing comma
+
+                let mut body = wrapper.clone().into_bytes();
+                body.splice(wrapper_splice..wrapper_splice, logs);
```

Comment:
Frankly, I thought this was a huge hack, but if you like it, I'm happy. Best would of course be some kind of JSON serializer that could do incremental work, but it's not obvious how that might work in practice.

I am of course curious what could be done to improve on it.

Reply:
Yeah, ive been mulling over this problem for a while since this is a major thing with making easy to write http sinks. I plan on seeing if I can clean up this hack next week. I think you have basically the gist of it. I think we could possibly write a custom body actually for http that could basically have slots for 3 slices so we don't have to do any double allocations like what is happening now. The other idea also is to use bytes as the "static" bytes and arc clone them rather then recopying the vec.

But I think for this right now, this is a good hack and its not messy either so I'm happy to keep this for now.